lib/pscom4ivshmem/direct_mode/psivshmem_malloc.c

```diff
@@ -0,0 +1,197 @@
+/*
+ * ParaStation
+ *
+ * Copyright (C) 2013 ParTec Cluster Competence Center GmbH, Munich
+ *
+ * All rights reserved.
+ *
+ * Author:	Jens Hauke <hauke@par-tec.com>
+ */
+
+#include <sys/ipc.h>
+#include <sys/shm.h>
+#include <assert.h>
+#include <stdlib.h>
+#include <errno.h>
+#include <malloc.h>
+#include <string.h>
+#include <stddef.h>
+
+#include "psivshmem_malloc.h"
+#include "../metadata.h"
+#include "pscom_env.h"
+//#include "../pscom4ivshmem/psivshmem.h"   //allows to handle pci device! //double include
+#include "../pscom_ivshmem.h"
+
+struct Psivshmem psivshmem_direct_info = {
+	.base = NULL,
+	.baseoffset = 0,
+	.tail = NULL,
+	.size = 0,
+//	.ivshmemid = 0,
+	.msg = "libpsivshmem_malloc.so not linked.",
+};
+
+
+struct Psivshmem_config {
+	size_t	min_size;
+	size_t	max_size;
+};
+
+
+static
+struct Psivshmem_config psivshmem_config = {
+	.min_size = 32UL * 1024 * 1024 /* 32MiB */,
+	.max_size = 64UL *1024 * 1024, // 64MiB     * 1024, /* 64 GiB */
+
+};
+
+/* Initialize base pointer with a shared mem segment. Return 0 on success, -1 else */
+static
+int psivshmem_init_base(void)
+{
+//	int ivshmemid;
+	void *buf;
+	size_t size = psivshmem_config.max_size;
+
+	ivshmem_conn_t ivshmem;
+	
+	// Warning Fix: 
+	ivshmem.device = NULL;
+	// ToDo 
+	
+	ivshmem_pci_dev_t* device_ptr = ivshmem.device;
+	
+	psivshmem_init_uio_device(device_ptr); //init device
+
+/*
+	buf = psivshmem_alloc_memory(&ivshmem->device, sizeof(psivshmem_com_t)); //returns ptr to first byte or NULL on error  
+
+	memset(buf, 0, sizeof(psivshmem_com_t));  // init with zeros
+*/
+
+	while (1) {
+	//	ivshmemid = shmget(/*key*/0, size,  /*SHM_HUGETLB |*/ SHM_NORESERVE | IPC_CREAT | 0777);
+	
+		buf = psivshmem_alloc_mem(device_ptr,size); //returns ptr to first byte or NULL on error  
+
+		if (buf != 0) break; // success with size bytes
+	//	if (errno != ENOSPC && errno != EINVAL) goto err; // error, but not "No space left on device" or EINVAL
+		size = size * 3 / 4; // reduce allocated size
+		if (size < psivshmem_config.min_size) break;
+	}
+	if (buf == 0 ) goto err;
+
+//	buf = shmat(ivshmemid, 0, 0 /*SHM_RDONLY*/);
+
+
+//	if (((long)buf == -1) || !buf) goto err_shmat;
+
+//	shmctl(ivshmemid, IPC_RMID, NULL); /* remove ivshmemid after usage */   //UNSOLVED: How to remove mem region after usage? (clear bit in bitmap)!!
```

Comment:
see above comment concerning unneeded code.

Is that issue solved already? Otherwise, place a TODO including a short problem description.